File: rankedResult.java
package com.contentguard.model;
import java.time.LocalDateTime;
import java.util.List;
public class RankedResult implements Comparable<RankedResult> {
    private String submissionId;
    private double aggregatedScore;
    private String plagiarismLabel;
    private int rank;
    private List<RawMatch> topMatches;
    private LocalDateTime processedAt;
    public RankedResult() {}
    public RankedResult(String submissionId,
                        double aggregatedScore,
                        String plagiarismLabel,
                        List<RawMatch> topMatches) {
        this.submissionId = submissionId;
        this.aggregatedScore = aggregatedScore;
        this.plagiarismLabel = plagiarismLabel;
        this.topMatches = topMatches;
        this.processedAt = LocalDateTime.now();
    }
    @Override
    public int compareTo(RankedResult other) {
        int cmp = Double.compare(other.aggregatedScore, this.aggregatedScore);
        if (cmp == 0) {
            return this.submissionId.compareTo(other.submissionId);
        }
        return cmp;
    }
    public String getSubmissionId() {
        return submissionId;
    }
    public void setSubmissionId(String submissionId) {
        this.submissionId = submissionId;
    }
    public double getAggregatedScore() {
        return aggregatedScore;
    }
    public void setAggregatedScore(double aggregatedScore) {
        if (aggregatedScore < 0 || aggregatedScore > 100) {
            throw new IllegalArgumentException("Aggregated score must be between 0 and 100");
        }
        this.aggregatedScore = aggregatedScore;
    }
    public String getPlagiarismLabel() {
        return plagiarismLabel;
    }
    public void setPlagiarismLabel(String plagiarismLabel) {
        this.plagiarismLabel = plagiarismLabel;
    }
    public int getRank() {
        return rank;
    }
    public void setRank(int rank) {
        this.rank = rank;
    }
    public List<RawMatch> getTopMatches() {
        return topMatches;
    }
    public void setTopMatches(List<RawMatch> topMatches) {
        this.topMatches = topMatches;
    }
    public LocalDateTime getProcessedAt() {
        return processedAt;
    }
    public void setProcessedAt(LocalDateTime processedAt) {
        this.processedAt = processedAt;
    }
    @Override
    public String toString() {
        return String.format(
                "RankedResult[rank=%d, submission=%s, score=%.2f%%, label=%s]",
                rank,
                submissionId,
                aggregatedScore,
                plagiarismLabel
        );
    }
}

File: rankedResult.cpp
#ifndef RANKEDRESULT_HPP
#define RANKEDRESULT_HPP
#include <string>
#include <vector>
#include "RawMatch.hpp"
class RankedResult {
private:
    string submissionId;
    double aggregatedScore;
    string plagiarismLabel;
    int rank;
    vector<RawMatch> topMatches;
public:
    RankedResult(
        const string& submissionId,
        double aggregatedScore,
        const string& plagiarismLabel,
        const vector<RawMatch>& topMatches
    ) : submissionId(submissionId),
        aggregatedScore(aggregatedScore),
        plagiarismLabel(plagiarismLabel),
        topMatches(topMatches),
        rank(0) {}
    bool operator<(const RankedResult& other) const {
        return aggregatedScore < other.aggregatedScore;
    }
    void setRank(int r) {
        rank = r;
    }
    int getRank() const {
        return rank;
    }
    double getAggregatedScore() const {
        return aggregatedScore;
    }
    string getSubmissionId() const {
        return submissionId;
    }
    string getPlagiarismLabel() const {
        return plagiarismLabel;
    }
};
#endif
